### ISSUE TITLE
SEAD / SAMs

### ISSUE THREAD

**Issue by wrycu** [shdwp/dcs_liberation#61]
Now that we've got player-controllable SEAD-capable aircraft for both sides, can we get some actual SAMs defending airbases? The most lethal I've seen so far is very light AAA or a very limited number of IR SAMs.  Is it possible to add actual radar SAMs, and/or more than a single air defense unit protecting airbases?

Seems a bit misleading to call the mission type "Strike/SEAD" when there appears to never be anything to suppress.  It also makes working toward capturing an airbase feel like a chore, as there are clearly no air defenses at it at any given time (in the intelligence view - I understand there are unit caps and you may not want to spawn all of the defenses for every mission.)

**Comment by shdwp** [comment]
There should be a sufficient number of SAM strike targets during Strikes, and during base attacks base would spawn additional SAMs (provided that the base have one; usually if it goes down to zero strength it doesn't).

Types of vehicles should also scale with airbases, first few targets will only include Shilkas, but then Strela's will start to appear as well as Osa's). I recommend just skimming trough first few bases if you find them too easy.

**Comment by wrycu** [comment]
I'm talking about SAMs more advanced than SA-8's. Give us advanced SAM sites!

(as a side note, I've done a terrible job of implementing this on my forked copy and you're welcome to adapt code to this version)

https://github.com/wrycu/dcs_liberation/commit/36aa4edb059993529bad9667284e9915784fd22f
https://github.com/wrycu/dcs_liberation/commit/89f122c3257bb89e166b71ea742fdfebcf2bb372

**Comment by Khopa** [comment]
@wrycu Tested your code and got to attack an SA-3 Site at an air defense site strike location. Site was nicely set up, and fired at me.

Good Job :+1: 

I destroyed all the units in the group, and they appeared in the debriefing as destroyed, but then the air defense site was still considered active on the map. 

I've been a bit out of loop when it come to dcs liberation lately, so i'm not exactly sure if it's a bug or a feature.  (Maybe it's normal, as the airbase linked to the site still had other air defense units in it's inventory ?)

**Comment by wrycu** [comment]
That's a bug. Any chance you could post the debrief log so I can try and squash it?

**Comment by Khopa** [comment]
Well, I don't have the file anymore. But i will try to reproduce as soon as i can.

**Comment by Khopa** [comment]
So here is the thing : 

Running a strike mission from Al Dhafra to Al Maktoum on PG map; 
Al Maktoum has one S-125 system (SA-3)

![image](https://user-images.githubusercontent.com/2546901/60389019-6f18c680-9aba-11e9-8669-c5cf293ce1f6.png)

Mission that is generated in DCS : 

[liberation_nextturn_quick.zip](https://github.com/shdwp/dcs_liberation/files/3342318/liberation_nextturn_quick.zip)

![image](https://user-images.githubusercontent.com/2546901/60389034-c9198c00-9aba-11e9-8eb2-090f06a2208d.png)

Corresponding to these targets : 

![image](https://user-images.githubusercontent.com/2546901/60389046-054cec80-9abb-11e9-95e7-dc65ea4e75ed.png)

(Note that, 2 SA-3 site have been generated, even though there is only one system in inventory. But i 'm not sure that the air defense site generation is linked to the airbase inventory, so maybe not really an issue, but it's still a bit confusing if this is correct by design.)

Then I launch the mission and destroy the AA1 site, dropping 2xMk84 on the radars and then finished the launchers with guns : 

![image](https://giant.gfycat.com/WelldocumentedFavorableHydra.gif)

And in DCS Liberation we can see that everything is considered destroyed : 

![image](https://user-images.githubusercontent.com/2546901/60389210-13e8d300-9abe-11e9-831f-6443a3453a6f.png)

I press "Okay".
Al Maktoum has lost a bit of power, but the air defense site is still present on map :

![image](https://user-images.githubusercontent.com/2546901/60389314-accc1e00-9abf-11e9-8c2c-a8818f7a69f5.png)

Tried to check if that was an issue with the map refresh, so i restarted DCS Liberation after that, but it was the same; 

Also note that though, 2 Mig-23MLD were destroyed according to the briefing, there are still accounted for on the following turn in Al Maktoum inventory. The M-2000C count was correctly updated though. 

***

And here is the debriefing file : 

[test.log](https://github.com/shdwp/dcs_liberation/files/3342312/test.log)

The interesting parts being : 

```
              [...]

              [34] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "42",
			["t"] = 57757.561,
			["type"] = "dead",
		},
                
               [...]

                [36] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "43",
			["t"] = 57785.764,
			["type"] = "dead",
		},

               [...]

               [55] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "45",
			["t"] = 57817.342,
			["type"] = "dead",
		},
		[56] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "44",
			["t"] = 57823.638,
			["type"] = "dead",
		},

               [...]

               [75] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "46",
			["t"] = 57858.915,
			["type"] = "dead",
		 },

               [...]

               [84] = {
			["initiator"] = "aa|6|24|1 Unit #1",
			["initiatorMissionID"] = "47",
			["t"] = 57916.867,
			["type"] = "dead",
		},

```

The thing i find weird, is that it's always "aa|6|24|1 Unit # 1" in the events
So i checked in the miz file, and all the units in the group do indeed have the same name : "aa|6|24|1 Unit # 1".

This is something you can't do in the mission editor usually, the editor UI would prevent you to give the same name to 2 units.

So this might be the issue, but yet the debriefing parser seems to correctly identify the destroyed unit type.

Note : I might have a look at the debriefing code myself later.

**Comment by wrycu** [comment]
Yeah, I didn't consider the "accounting for things being destroyed" when implementing the sites. I'm working on that now.  One to note is that I permitted AA defense sites to have randomly selected anti-air composition (without regard for what the base itself has present) in order to make things more interesting.

I see this as two problems - one that the air defense site isn't tracked as destroyed (which I haven't yet begun to look into) and one that SAM sites defending a base are not counted as destroyed (which I am test a fix for.)

Thank you for the detailed write-up and inclusion of files, it will assist greatly in finding the issue without having to fly it myself.

**Comment by Khopa** [comment]
Yes, i've also been working on some extensive changes on my own branch :
https://github.com/Khopa/dcs_liberation/tree/develop_khopa

* The possibility to select a player faction and an ennemy faction, so that it is possible to play with different units sets and era. 
* A turn counter (with date, and time of day)

I also tried a quick fix at the aa gen code, so that the unit name is correctly 'numbered', but it didn't seem to fix the issue. But I had a lot of changes on my branch done by then, so I might have broken some stuff. 

***

Small other issues and thoughts related to the air defense : 

* **Remove the AA units from the airbase inventory :** Even if the enemy has 10 aa units in it's inventory, he will only deploy as much as the number of aa-site he has. To improve things, I think we should change the initial generation so that when an enemy aa-site is generated at the game startup, the units placed in it will also be generated at this moment and placed in the site; so they would not be managed in the airbase inventory anymore. The enemy air defense network and the overall Electronic Warfare environment can remain consistent and logical accross each missions played, as the deployed units would always remain the same in each site. 

* **AA threat always there :** I think the AA Defense site should always be there in the generated missions (at least, for '2/3' bases away from the conflict zone) regardless of the mission type chosen. Currently when you select a "frontline assault" mission, where you have to run CAS, there is no AA defense generated. This would also allow the EW environment to remain consistent, and make sure that the SAMs remains a constant and real threat to the player, so that running SEAD/DEAD mission is actually important when the enemy.

* **Close range AA on the frontline :** Some AA systems are meant to be on the frontline to provide close range protection to the ground units. Say manpads and others Infrared SAM, as well as AAA. They should be generated on the frontline with the ground units, a few hundred meters in the back.  

* **Player's SAM :** Currently, it is not really possible for the player to setup it's own air defense system.

**Comment by wrycu** [comment]
I've got a fix working for the second issue I outlined (https://github.com/wrycu/dcs_liberation/commit/b1576e8f15eb78253cef83b8121e9e352d810819) and am still working on what you outlined about AA defense sites not being properly removed. 
***
I took a gander at your code and saw you pulled in my entire fork. I wanted to let you know that I'd made other changes in my branch, to avoid any nasty surprises. My changes include:
* The enemy gains income and can perform actions on every turn, not just those when the player performs an action
* Raised the AA defense limit for bases
* Tweaked air defense costs
* Attacking a base now results in it launching all CAP-capable aircraft to defend itself
* Each enemy CAP flight spawns at a random time during the mission instead of at the start
* FARPs and other strike targets have light AA defending them (which is closely related to your third bullet)
* Pushed base defense sites further away from the base (to account for longer range units)

Overall, I like what you're doing (the turn counter _could_ even lead to aircraft becoming available as certain dates are hit.)
***
As to your thoughts on the AA stuff - 
* **Remove the AA units from the airbase inventory** - love this. I didn't want to get too into the liberation-specific code in my tweaks, but I think this would be a very much welcome change. Long-term, this could permit adding in IADS functionality and intelligently mixing units of various types to provide better coverage
* **AA threat always there** - totally agree. This was actually on my (WIP) to-do list
![image](https://i.imgur.com/DF35ZZE.png)
* **Close range AA on the frontline** - again, totally agree. I've already begun work on this, to some degree, by adding it to FARPs
* **Player's SAM** - not sure how strongly I feel about this one, but sure. If the enemy will actually make it worthwhile, then sounds good
***
Overall it seems like our plans align pretty closely. It might be worth merging our efforts and collaborating on one repo. Just a thought.

**Comment by Khopa** [comment]
> I wanted to let you know that I'd made other changes in my branch, to avoid any nasty surprises. 

Ok, I will try to merge it into my branch.

> Each enemy CAP flight spawns at a random time during the mission instead of at the start

Nice, i wanted to do something like this as well. Something like planning enemy CAP and CAS strike for the next 2 hours or so as to give the mission a dynamic feel; instead of the single furball that every CAP mission used to consist of. 

Also it would be nice to have a script, to make the enemy scramble interceptors when you get too close from an airbase or an objective. Either we could just use pydcs to generate the triggers that would do that, or we could generate lua code to embed in the mission to do it.

> FARPs and other strike targets have light AA defending them (which is closely related to your third bullet)

Agreed as well. 

***

> Player's SAM - not sure how strongly I feel about this one, but sure. If the enemy will actually make it worthwhile, then sounds good

I think it's important, as it's always a good feeling when you're in danger to be able to hide behind your air defense network, it gives a feeling of safety when you're cruising in ally airspace. 

***

> Overall it seems like our plans align pretty closely. It might be worth merging our efforts and collaborating on one repo. Just a thought.

Agreed as well, we could setup a github organization and setup a new repo there, and invite the others contributors.

Might be worth setting up a Discord server as well, to discuss development, as well as to have a hub for the users community in the future; And it seems like your todo list screenshot is already coming from Discord ?

**Comment by wrycu** [comment]
I created a server so we can chat without clogging up the issues here.  Note that I'm UTC-7, so it may take an effort to get a real-time conversation going :).

https://discord.gg/7q8Xms

**Comment by wrycu** [comment]
I'm gonna close this as I implemented the tougher SAMs in f5882ccea570e670e6bd2d9016b24faced88663b.  A full rework of it has a different issue tracking progress.
